Intermittent `psycopg2.InterfaceError: cursor already closed` in tests

The links will not be useful  in a month or two, but here's a failed job on a PR: https://github.com/opendatacube/datacube-explorer/actions/runs/14381753644/job/40327020011, and here's the same job going green when retriggered: https://github.com/opendatacube/datacube-explorer/actions/runs/14381753644/job/40329616425

Here's what the log says:
```
________________________ test_stac_collection[default] _________________________

self = <sqlalchemy.engine.cursor.CursorFetchStrategy object at 0x7f992d6f4c00>
result = <sqlalchemy.engine.cursor.CursorResult object at 0x7f9920663af0>
dbapi_cursor = <cursor object at 0x7f9924aa9210; closed: -1>, hard_close = False

    def fetchone(
        self,
        result: CursorResult[Any],
        dbapi_cursor: DBAPICursor,
        hard_close: bool = False,
    ) -> Any:
        try:
>           row = dbapi_cursor.fetchone()
E           psycopg2.InterfaceError: cursor already closed

/usr/local/lib/python3.12/dist-packages/sqlalchemy/engine/cursor.py:1106: InterfaceError

The above exception was the direct cause of the following exception:

stac_client = <FlaskClient <Flask 'cubedash'>>

    @pytest.mark.parametrize("env_name", ("default",), indirect=True)
    def test_stac_collection(stac_client: FlaskClient):
        """
        Follow the links to the "high_tide_comp_20p" collection and ensure it includes
        all of our tests data.
        """
    
        collections = get_json(stac_client, "/stac")
        for link in collections["links"]:
            if link["rel"] == "child" and link["title"] == "high_tide_comp_20p":
                collection_href = link["href"]
                break
        else:
            raise AssertionError("high_tide_comp_20p not found in collection list")
    
        scene_collection = get_collection(stac_client, collection_href, validate=False)
        # HACK: Heavy handed bypass for pytest approximate
        import copy
    
        proxy_scene_collection = copy.deepcopy(scene_collection)
        proxy_scene_collection["extent"]["spatial"]["bbox"][0] = [
            pytest.approx(p, abs=0.001)
            for p in scene_collection["extent"]["spatial"]["bbox"][0]
        ]
    
        assert proxy_scene_collection == {
            "stac_version": "1.1.0",
            "type": "Collection",
            "id": "high_tide_comp_20p",
            "title": "high_tide_comp_20p",
            "license": "CC-BY-4.0",
            "description": "High Tide 20 percentage composites for entire coastline",
            "extent": {
                "spatial": {
                    "bbox": [
                        [
                            pytest.approx(112.223_058_990_767_51, abs=0.001),
                            pytest.approx(-43.829_196_553_065_4, abs=0.001),
                            pytest.approx(153.985_054_424_922_77, abs=0.001),
                            pytest.approx(-10.237_104_814_250_783, abs=0.001),
                        ]
                    ]
                },
                "temporal": {
                    "interval": [["2008-06-01T00:00:00Z", "2008-06-01T00:00:00Z"]]
                },
            },
            "links": [
                {
                    "rel": "root",
                    "href": "http://localhost/stac",
                    "type": "application/json",
                    "title": "Default ODC Explorer instance",
                },
                {
                    "rel": "self",
                    "href": stac_url("collections/high_tide_comp_20p"),
                },
                {
                    "rel": "items",
                    "href": stac_url("collections/high_tide_comp_20p/items"),
                },
                {
                    "rel": "http://www.opengis.net/def/rel/ogc/1.0/queryables",
                    "href": stac_url("collections/high_tide_comp_20p/queryables"),
                },
                {
                    "rel": "child",
                    "href": stac_url("catalogs/high_tide_comp_20p/2008-6"),
                },
            ],
            # "providers": [], // FIXME: These disappeared somewhere along the way ?
            # "stac_extensions": [],
        }
        # HACK: Make things float again
        assert_collection(scene_collection)
        for link in scene_collection["links"]:
            if link["rel"] == "items":
                item_links = link["href"]
                break
>       validate_items(_iter_items_across_pages(stac_client, item_links), expect_count=306)

integration_tests/test_stac.py:722: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
integration_tests/test_stac.py:343: in validate_items
    for item in items:
integration_tests/test_stac.py:273: in _iter_items_across_pages
    items = get_items(client, url)
integration_tests/test_stac.py:241: in get_items
    data = get_geojson(client, url)
integration_tests/asserts.py:77: in get_geojson
    data = get_json(client, url)
integration_tests/asserts.py:98: in get_json
    rv: Response = client.get(url, follow_redirects=True)
/usr/local/lib/python3.12/dist-packages/werkzeug/test.py:1162: in get
    return self.open(*args, **kw)
/usr/local/lib/python3.12/dist-packages/flask/testing.py:234: in open
    response = super().open(
/usr/local/lib/python3.12/dist-packages/werkzeug/test.py:1116: in open
    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)
/usr/local/lib/python3.12/dist-packages/werkzeug/test.py:988: in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
/usr/local/lib/python3.12/dist-packages/werkzeug/test.py:1264: in run_wsgi_app
    app_rv = app(environ, start_response)
/usr/local/lib/python3.12/dist-packages/flask/app.py:1536: in __call__
    return self.wsgi_app(environ, start_response)
/usr/local/lib/python3.12/dist-packages/werkzeug/middleware/proxy_fix.py:183: in __call__
    return self.app(environ, start_response)
/usr/local/lib/python3.12/dist-packages/flask/app.py:1514: in wsgi_app
    response = self.handle_exception(e)
/usr/local/lib/python3.12/dist-packages/flask_cors/extension.py:176: in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
/usr/local/lib/python3.12/dist-packages/flask/app.py:1511: in wsgi_app
    response = self.full_dispatch_request()
/usr/local/lib/python3.12/dist-packages/flask/app.py:919: in full_dispatch_request
    rv = self.handle_user_exception(e)
/usr/local/lib/python3.12/dist-packages/flask_cors/extension.py:176: in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
/usr/local/lib/python3.12/dist-packages/flask/app.py:917: in full_dispatch_request
    rv = self.dispatch_request()
/usr/local/lib/python3.12/dist-packages/flask/app.py:902: in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
cubedash/_stac.py:1070: in stac_search
    _handle_search_request(request.method, args, products)
cubedash/_stac.py:618: in _handle_search_request
    feature_collection = search_stac_items(
cubedash/_stac.py:774: in search_stac_items
    items = list(
cubedash/summary/_stores.py:1154: in search_items
    for r in self.e_index.execute_query(query):
/usr/local/lib/python3.12/dist-packages/sqlalchemy/engine/result.py:527: in iterrows
    for raw_row in self._fetchiter_impl():
/usr/local/lib/python3.12/dist-packages/sqlalchemy/engine/cursor.py:2126: in _fetchiter_impl
    row = fetchone(self, self.cursor)
/usr/local/lib/python3.12/dist-packages/sqlalchemy/engine/cursor.py:1111: in fetchone
    self.handle_exception(result, dbapi_cursor, e)
/usr/local/lib/python3.12/dist-packages/sqlalchemy/engine/cursor.py:1082: in handle_exception
    result.connection._handle_dbapi_exception(
/usr/local/lib/python3.12/dist-packages/sqlalchemy/engine/base.py:2352: in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sqlalchemy.engine.cursor.CursorFetchStrategy object at 0x7f992d6f4c00>
result = <sqlalchemy.engine.cursor.CursorResult object at 0x7f9920663af0>
dbapi_cursor = <cursor object at 0x7f9924aa9210; closed: -1>, hard_close = False

    def fetchone(
        self,
        result: CursorResult[Any],
        dbapi_cursor: DBAPICursor,
        hard_close: bool = False,
    ) -> Any:
        try:
>           row = dbapi_cursor.fetchone()
E           sqlalchemy.exc.InterfaceError: (psycopg2.InterfaceError) cursor already closed
E           (Background on this error at: https://sqlalche.me/e/20/rvf5)

/usr/local/lib/python3.12/dist-packages/sqlalchemy/engine/cursor.py:1106: InterfaceError
```

Is anything special done in the tests, or is this likely to be a problem in production too? (Not that I want flaky tests, but I prefer tests crashing intermittently over production services crashing intermittently.)

--- a/cubedash/index/api.py
+++ b/cubedash/index/api.py
@@ -29,9 +29,18 @@ class ExplorerAbstractIndex(ABC):
         self.engine = index._db._engine  # type: ignore[attr-defined]
 
     # need to add an odc_index accessor
-    def execute_query(self, query):
+    def execute_query(self, query) -> list[Row]:
+        with self.engine.connect() as conn:
+            return conn.execute(query).fetchall()
+
+    def execute_query_scalar(self, query) -> Any:
+        with self.engine.connect() as conn:
+            return conn.execute(query).scalar()
+
+    def execute_ddl(self, query) -> int:
         with self.engine.begin() as conn:
-            return conn.execute(query)
+            results = conn.execute(query)
+            return results.rowcount
 
     def make_dataset(self, row):
         # pylint: disable=protected-access

--- a/cubedash/summary/_stores.py
+++ b/cubedash/summary/_stores.py
@@ -609,7 +609,7 @@ class SummaryStore:
         """
         Drop all cubedash-specific tables/schema.
         """
-        self.e_index.execute_query(
+        self.e_index.execute_ddl(
             DDL(f"drop schema if exists {_schema.CUBEDASH_SCHEMA} cascade")
         )
 
@@ -1043,7 +1043,7 @@ class SummaryStore:
             query = self._add_filter_to_query(
                 query, field_exprs, filter_lang, filter_cql
             )
-        result = self.e_index.execute_query(query).fetchall()
+        result = self.e_index.execute_query(query)
 
         if len(result) != 0:
             return result[0][0]
@@ -1384,7 +1384,7 @@ class SummaryStore:
         Any change timestamps stored in the database are using database-local
         time, which could be different to the time on this current machine!
         """
-        return self.e_index.execute_query(select(func.now())).scalar()
+        return self.e_index.execute_query_scalar(select(func.now()))
 
     def _newest_known_dataset_addition_time(self, product_name: str) -> datetime | None:
         """

